### PR TITLE
Add support for keccak256 hashing

### PR DESCRIPTION
**Pull request by bbondy**
Resolves https://github.com/brave/brave-browser/issues/14329

Ethereum uses the Keccak256 hashing function in many places. This differs from SHA3 by some parameters and Chromium doesn't have support for it. So this issue is to track the inclusion of the ethash hashing library and related work.

It is easiest to review this PR 1 commit at a time.

## Submitter Checklist:

- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)
- [ ] Requested a security/privacy review as needed

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Review by orspetol — APPROVED** [review]
Thank you for validating both empty and non-empty input in the test suite. LGTM.

**Comment by orspetol** [comment]
@bbondy Do you think we need an explainer in the code about keccak vs FIPS-202? Totally a nit but I am curious what you think.

**Comment by bbondy** [comment]
> @bbondy Do you think we need an explainer in the code about keccak vs FIPS-202? Totally a nit but I am curious what you think.

I think maybe more for a wiki page that describes this if we do. Because if someone is looking at an Ethereum implementation they'd expect this, so it's not out of the ordinary.

**Review by SergeyZhukovsky — APPROVED** [review]
++

File: third_party/ethash/keccak_override.c [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#include "build/build_config.h"
+
+// The use of __builtin_cpu_init is wrapped around an __X86_64__ ifdef.
+// Since this isn't available in vc++ and since it's only needd on macOS,
+// we exclude this block with the following hack.
```
Comment (SergeyZhukovsky) [inline third_party/ethash/keccak_override.c:5]: ++

**Review by SergeyZhukovsky — APPROVED** [review]
++

**Pull request merged**